### ISSUE TITLE
Add get_player_velocity() method

### ISSUE THREAD

**Issue by 4Evergreen4** [minetest/minetest#1176]
I ran into a case which required checking a given player's velocity while modding, so I thought I would ask for it here.

**Comment by kwolekr** [comment]
Hi Eliivention, was just wondering if you would share your use case for this function.

**Comment by 4aiman** [comment]
@Elinvention while it's *absolutely wonderful* that you've added this function, the way you did it make mod creators to do unnecessary work.

Entities have get_velocity() method already.
It would be much easier to write
```
local vel = 0
if obj:get_velocity then 
   vel = obj:get_velocity()
end
```
than
```
local vel = 0
if obj:is_player() then
   if obj:get_player_velocity then
      vel = obj:get_player_velocity()
   end
else
   vel = obj:get_velocity()
end
```

Now add the fact that the ```is_player``` function was non-existent for non-players prior to ~month ago:
```
local vel = 0
if obj:is_player and obj:is_player() then
   if obj:get_player_velocity then
      vel = obj:get_player_velocity()
   end
else
   vel = obj:get_velocity()
end
```
*Note*: checks are necessary to support the previous versions of Minetest.

TL;DR
I'd suggest renaming the method to correspond that of the non-player entities.

**Comment by MirceaKitsune** [comment]
This is something which we really, really need in my opinion. Not having a way to get the player's velocity at this stage of the project is simply silly.

I am however aware of why it's the case: Players use a different physics system from Lua entities. Which I never really understood, and I think is itself a problem that can preferably be solved someday. Whether it's with the current player physics or not however, a way to get the player's velocity is an essential part of mod making... obviously without having to manually compare position changes.

**Comment by Elinvention** [comment]
Well, I just wanted to make the player scream while falling from high place. That's how I found out there was no get_velocity for players. I had to compute the average speed every second and this led to out of sync screams.
@4aiman I see the inconvenience so I am going to rename it as you suggested.

**Comment by est31** [comment]
@4aiman , what would be a usecase for generic velocity retrieving code? we have "make player scream when falling" -- what do we need else?

**Comment by est31** [comment]
Also, the code can be simplified:
```Lua
local vel = 0
if obj:get_player_velocity then
      vel = obj:get_player_velocity()
elseif obj:get_velocity then
   vel = obj:get_velocity()
end
```

**Comment by est31** [comment]
I think screaming falling players is a good usecase

**Comment by C1ffisme** [comment]
You aren't supposed to ask "What's the use of this", because when you think for about 5 seconds, you can't think of anything. But when you are building a big mod project and you need it, then you can use it.

**Comment by 4aiman** [comment]
@est31 

> Also, the code can be simplified:

Holy cow! Easy and elegant :)


What's the "usecase" you ask?..
The function is already there.
So, the usecase of a *generic* method is only one - eliminate unnecessary checks. Even the simplified version of the code above can be simplified further with the generic method name.


As for the ability to get player speed itself...
The following does not depend on the name of a method, but having a generic one is preferable - less checks == faster code:

- I can add some critical damage if the speed is above normal. 
(Right now all I can do is to depend onto physics_override multiplier. I had to abandon the idea of speed-dependent critical strikes for mobs (too laggy).
In the current situation striking a critical shot while jumping is an impossible task for both players and mobs.)

- The other use may be lowering damage for a moving player or entity (i.e when either one receives damage).

- With camera position settable (that is done, right?), a speedometer for vehicles can be shown easily.

- Jump multiplier may be altered when a player or entity runs fast. The faster - the higher.

**Comment by mrob27** [comment]
MirceaKitsune commented, *"I am however aware of why it's the case: Players use a different physics system from Lua entities. Which I never really understood, and I think is itself a problem that can preferably be solved someday. …"*

As I understand it, each client's camera position is tied to the position of the local player on that client, and that player's physics are computed in the client, and pushed to the other clients via the server. Whereas the physics for non-player entities are computed in the server, and pushed to all clients.

If the players' physics were all done on the server, there would be lagging in the camera position when you walk up stairs, hit walls or do other things where the camera position is affected by collision-related physics.

A way to see this lagginess in action is to **set_attach()** yourelf to an object, such as a boat. Get in a boat and move as fast as you can. The client has to keep updating your position (and thus the camera position) to agree with the position of the boat, which it is getting from the server. Even in singleplayer mode, the animation will occasionally stutter. (It may help to temporarily modify your minetest_game/mods/boats/init.lua and increase the boat speed limit, normally 4.5. Doubling it to 9.0 was enough for me to see the stuttering.)

**Comment by est31** [comment]
You are completely right, this is precisely the reason why any server side player physics should be avoided. That "problem" MirceaKitsune sees can't be "solved", as its no problem at all. We can perhaps add cheat prevention for things like noclip or fly, and perhaps even calculate things like falling damage on the server, but it will never be the same as normal lua entities, which are fully server side.

If clientside scripting arrives, custom physics can indeed be introduced, but on the client. We won't be abled to stop modders from shooting themselves and users into the foot by relaying the client side API to the server side, but that's their fault then, and its harder to do than keeping the calculations where they belong.

**Comment by est31** [comment]
Unfortunately there aren't much docs I could point to, but if you want to read some docs, consider the `doc/` subdir in the engine source code, and the wiki at http://dev.minetest.net . Note that they might be outdated in smaller details. You know, like docs always are :)

If you have questions, you can ask them on irc: http://dev.minetest.net/IRC (devs are in #minetest-dev)

**Comment by mrob27** [comment]
Thanks for confirming, **est31**. I believe that's an essential part of any multi-player game design: Compute each player's physics locally on their client, but do everyting else on a central server, both to close cheating loopholes, and so that players don't have to keep their software version and add-ons (mods) in sync with the server they're playing on.

Thanks for reminding me to look in the **doc** directory. Out-of-date is okay, I understand such things.

For now I'd rather avoid bothering the devs. I feel like I haven't paid my dues yet. I am partway through a painful process of getting Minetest to build entirely stand-alone (using no globally installed packages, because I need to do everything without invoking "root" or "sudo" permissions). That means doing something similar to a Windows build from scratch — but I'm on a Mac with Mac OS 10.6.8, and I can't update the OS.

Irrlight and its demos build fine, which required finding and patching a couple lines of source code. Minetest compiles but naturally it doesn't link. I have to do similar work on the other essential components, like SQL. After getting a working Minetest built from source I'll feel qualified to reverse-engineer a description of its design architecture :D or to ask questions.
